# **VERSUSVIRUS - Team 306 - Real Time Bed Monitoring**

---



## **HOW LONG WILL A PATIENT STAY HOSPITALIZED?**

Regression on the duration of stay using an AutoML & OLS approach.

Data has been previously synthetized and designed to represent the distribution of Covid-19 patients admitted to Swiss hospitals and their medical conditions

In [0]:
# Load the Drive helper and mount
from google.colab import drive
import os
import pandas as pd
!pip install tpot
import tpot


drive.mount('/content/drive')
os.chdir("drive/My Drive/VersusVirus_Hackathon")

## Data Preparation

In [0]:
df = pd.read_excel('./TRAINING_DURATION_OF_STAY.xlsx')
df = pd.get_dummies(df, columns=['Admission Type', 'Gender'])

In [61]:
df.sample(5)

Unnamed: 0  Age  ... Gender_Female Gender_Male
7577        7577   40  ...             1           0
4304        4304   52  ...             0           1
9265        9265   29  ...             0           1
1111        1111   28  ...             0           1
6685        6685   75  ...             0           1

[5 rows x 60 columns]

## Define predictors and target variable

In [0]:
predictors = ['Age', 'Hypertension', 'Cardiovascular Disease', 'Diabetes',
       'Malignacy', 'Cerebrovascular disease', 'COPD',
       'Chronoc kidney disease', 'Chronic liver disease', 'HIV infection',
       'Fever', 'Fatigue', 'Dry cough ', 'Anorexia', 'Myalgia', 'Dyspnea',
       'Expectoration', 'Pharyngalgia', 'Diarrhea', 'Nausea', 'Dizziness',
       'Headache', 'Vomiting ', 'Abdominal Pain', 'Heart Rate (median)',
       'Respiratory Rate (median)', 'Mean arterial pressure  (median)',
       '# of pre_conditions', '# of symptoms', 'Admission Type_ICU - Invasive',
       'Admission Type_ICU - Non Invasive', 'Admission Type_Non-ICU',
       'Gender_Female', 'Gender_Male']

target = ['DUATION OF STAY']

In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    df[predictors], df[target].values[:, 0], test_size=0.25, random_state=42) #took 75% for training & 25% for testing

## Model Training

USE A TPOT PIPELINE TO DETERMINE BEST MODEL AMONGST MANY *(Running very long, adjust training parameters!)*

Evaluation Metric = RMSE. The error is calculated as the squareroot of the mean of the sum of all predicted duation of stay values - real duration of stay values for a given patient


$RMSE = \sqrt{\frac{1}{n}\Sigma_{i=1}^{n}{\Big(\frac{d_i -f_i}{\sigma_i}\Big)^2}}$

In [0]:
# First import the relevant TPOT pipeline
from tpot import TPOTRegressor 
tpot = TPOTRegressor()
# Fits the training data on the 
tpot.fit(X_train, y_train)
# Exports best pipeline to a python file so it can be used
tpot.export('./my pipeline.py')

In [70]:
import numpy as np
print("RMSE after training and model selection for best regressor: ",np.sqrt(abs(tpot.score(X_test, y_test))))

MESE after training and model selection for best regressor:  2.1333814787472622



FIT A BENCHMARK OLS MODEL

In [68]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np
ols1 = LinearRegression(normalize=True) #normalize inputs
ols1.fit(X_train, y_train)    # Fitting a linear model
y_pred = ols1.predict(X_test)
print("RMSE on test-set for OLS Benchmark Model",mean_squared_error(y_test, y_pred, squared=False))

RMSE on test-set for OLS Benchmark Model 2.1250818417065482


SCORE MODEL THAT CAME OUT OF THE TPOT PIPELINE

In [66]:
from sklearn.feature_selection import SelectPercentile, f_regression
from sklearn.linear_model import LassoLarsCV, RidgeCV
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline, make_union
from sklearn.preprocessing import Normalizer, RobustScaler
from sklearn.tree import DecisionTreeRegressor
from tpot.builtins import StackingEstimator

# Average CV score on the training set was: -4.811141924516791
exported_pipeline = make_pipeline(
    Normalizer(norm="l1"),
    SelectPercentile(score_func=f_regression, percentile=7),
    StackingEstimator(estimator=RidgeCV()),
    RobustScaler(),
    StackingEstimator(estimator=LassoLarsCV(normalize=False)),
    DecisionTreeRegressor(max_depth=3, min_samples_leaf=10, min_samples_split=7)
)

exported_pipeline.fit(X_train, y_train)
y_pred_automl = exported_pipeline.predict(X_test)
print("RMSE on test-set for AutoML best selected model: ",mean_squared_error(y_test, y_pred_automl, squared=False))

RMSE on test-set for AutoML best selected model:  2.1333814787472622


OLS MODEL SEEM TO BE ENOUGH

## OLS WITH VALIDATION SET

In [0]:
from sklearn.model_selection import train_test_split
df_sampled= df.sample(np.int(np.round(len(df)*0.95, 0))) # USE A RANDOM SUBSET AND LEAVE 5% AS VALIDATION SET 
df_val = df[~df.index.isin(df_sampled.index)]

X_train_, X_test_, y_train_, y_test_ = train_test_split(
    df_sampled[predictors], df_sampled[target].values[:, 0], test_size=0.25, random_state=42)

In [65]:
from sklearn.model_selection import cross_val_score
ols2 = LinearRegression(normalize=True)
ols2 = ols1.fit(X_train_, y_train_)
scores = cross_val_score(ols1, X_train_, y_train_, cv=5, scoring='neg_mean_squared_error')
print("Cross Validated - Test set validation score:" , np.sqrt(np.abs(scores.mean())))

Cross Validated - Test set validation score: 2.1724043554614143


RMSE on Validation Set

In [64]:
y_pred_val = ols2.predict(df_val[predictors])
print("RMSE on hold-out-set", mean_squared_error(df_val[target].values[:, 0], y_pred_val, squared=False)) #sqaured = False -> the error will not be squared to MSE hence, we ahave the RMSE

RMSE on hold-out-set 2.189016143081518


## Export Model for Deployment

In [0]:
import joblib 
# save the model to disk
filename = 'bed_duation_model_versus_virus_05APRIL2020.sav'
joblib.dump(ols2, filename)

In [0]:
#df_val[predictors].to_csv('./Deployment/INPUT_MODEL.csv')

**WE ARE ABLE TO PROVIDE A MODEL WITH A RMSE of ~2.2.**


